### ISSUE TITLE
OSX support

### ISSUE THREAD

**Issue by dbkaplun** [firehol/netdata#124]
Would be great to have OSX support for netdata as a lot of developers I know are switching from Linux to Mac for professional development. Here's how htop does it on Mac: https://github.com/hishamhm/htop/blob/master/darwin/DarwinProcessList.c

**Comment by ktsaou** [comment]
There is a PR to make it work on FreeBSD #187 
This might help in that direction.

**Comment by kfriede** [comment]
Releasing on it Homebrew at that point would be awesome too.

**Comment by hellobbn** [comment]
There is still an error: 
FreeBSD has a `thr.h` which the [PR](https://github.com/firehol/netdata/pull/187) uses to define `MADV_DONTFORK`
However in OSX it doesn't have the headfile.
So how can I solve it?
```
common.c: In function ‘mymmap’:
common.c:19:23: error: ‘INHERIT_NONE’ undeclared (first use in this function)
 #define MADV_DONTFORK INHERIT_NONE
                       ^

common.c:131:36: note: in expansion of macro ‘MADV_DONTFORK’
       int advise = MADV_SEQUENTIAL|MADV_DONTFORK;
                                    ^~~~~~~~~~~~~

common.c:19:23: note: each undeclared identifier is reported only once for each function it appears in
 #define MADV_DONTFORK INHERIT_NONE
                       ^

common.c:131:36: note: in expansion of macro ‘MADV_DONTFORK’
       int advise = MADV_SEQUENTIAL|MADV_DONTFORK;
                                    ^~~~~~~~~~~~~
```

**Comment by ktsaou** [comment]
compile with `-DMADV_DONTFORK=0` might make pass through this.

**Comment by hellobbn** [comment]
And After successfully compiling another error happened:
```
:-----------------------------------------------------------------------------
Running command:

groupadd -r netdata 

./netdata-installer.sh: line 234: groupadd: command not found
Adding netdata user account ...
./netdata-installer.sh: line 280: getent: command not found
:-----------------------------------------------------------------------------
Running command:

useradd -r -g netdata -c netdata -s /sbin/nologin -d / netdata 

./netdata-installer.sh: line 234: useradd: command not found
Fixing directory permissions for user netdata...
:-----------------------------------------------------------------------------
Running command:

chown -R netdata:netdata /usr/share/netdata/web 

chown: netdata: illegal group name
WARNING: Cannot change the ownership of the files in directory /usr/share/netdata/web to netdata...
:-----------------------------------------------------------------------------
Running command:

chmod 0775 /usr/share/netdata/web 

:-----------------------------------------------------------------------------
Running command:

chown -R netdata:netdata /etc/netdata 

chown: netdata: illegal group name
WARNING: Cannot change the ownership of the files in directory /etc/netdata to netdata...
:-----------------------------------------------------------------------------
Running command:

chmod 0775 /etc/netdata 

:-----------------------------------------------------------------------------
Running command:

chown -R netdata:netdata /var/cache/netdata 

chown: netdata: illegal group name
WARNING: Cannot change the ownership of the files in directory /var/cache/netdata to netdata...
:-----------------------------------------------------------------------------
Running command:

chmod 0775 /var/cache/netdata 

:-----------------------------------------------------------------------------
Running command:

chown -R netdata:netdata /var/log/netdata 

chown: netdata: illegal group name
WARNING: Cannot change the ownership of the files in directory /var/log/netdata to netdata...
:-----------------------------------------------------------------------------
```

also the error when starting netdata is:
```
16-04-17 07:16:08: ERROR: netdata: User netdata is not present. Ignoring option.
16-04-17 07:16:08: ERROR: netdata: Group netdata is not present. Ignoring option.
16-04-17 07:16:08: INFO: netdata: Listening on all IPs (IPv6 and IPv4)
16-04-17 07:16:08: ERROR: netdata: IPv6 bind() failed. Disabling IPv6. (errno 48, (null))
16-04-17 07:16:08: INFO: netdata: Listening on any IPs (IPv4).
16-04-17 07:16:08: ERROR: netdata: IPv4 bind() failed. (errno 48, (null))
16-04-17 07:16:08: FATAL: netdata: Cannot listen socket. # : Undefined error: 0
```

**Comment by ktsaou** [comment]
http://stackoverflow.com/questions/19071512/socket-error-errno-48-address-already-in-use

Address is already in use. Something else is listening on port 19999

**Comment by hellobbn** [comment]
print oul an error saying 
Access to file '/usr/share/netdata/web/' is not permitted.

already used `chown -R netdata:netdata /usr/share/netdata/web`
                     `chmod 0775 /usr/share/netdata/web`
sorry to disturb you.

**Comment by ktsaou** [comment]
no problem.

> 16-04-17 07:16:08: ERROR: netdata: User netdata is not present. Ignoring option.
> 16-04-17 07:16:08: ERROR: netdata: Group netdata is not present. Ignoring option.

have you added the user and group `netdata`?
after adding them, you should restart netdata.

Since netdata complains about the user/group missing, these files should be owned by the user/group netdata runs as. Can you find the user netdata runs as?

**Comment by hellobbn** [comment]
Now I have added them all and run the command 
`sudo netdata -p 65531 -u netdata`
it shows
`ERR_INCOMPLETE_CHUNKED_ENCODING`

**Comment by ktsaou** [comment]
edit the configuration and disable web compression (something is happening with zlib). It should be at the `[global]` section.

**Comment by ktsaou** [comment]
I guessed you receive this error on your browser.

If you don't have /etc/netdata/netdata.conf, this is what you need:

```
[global]
	enable web responses gzip compression = no
```

**Comment by hellobbn** [comment]
I wondering if this log can help:
```
16-04-17 11:38:16: INFO: netdata: Successfully became user 'netdata'.
16-04-17 11:38:16: INFO: netdata: NetData started on pid 4093
16-04-17 11:38:16: INFO: netdata: Starting thread tc.
16-04-17 11:38:16: INFO: netdata: Starting thread idlejitter.
16-04-17 11:38:16: INFO: netdata: TC thread created with task id -1
16-04-17 11:38:16: INFO: netdata: CPU Idle Jitter thread created with task id -1
16-04-17 11:38:16: INFO: netdata: Starting thread proc.
16-04-17 11:38:16: INFO: netdata: PROC Plugin thread created with task id -1
16-04-17 11:39:42: INFO: netdata: Successfully became user 'netdata'.
16-04-17 11:39:42: INFO: netdata: NetData started on pid 6334
16-04-17 11:39:42: INFO: netdata: Starting thread tc.
16-04-17 11:39:42: INFO: netdata: Starting thread idlejitter.
16-04-17 11:39:42: INFO: netdata: TC thread created with task id -1
16-04-17 11:39:42: INFO: netdata: CPU Idle Jitter thread created with task id -1
16-04-17 11:39:42: INFO: netdata: Starting thread proc.
16-04-17 11:39:42: INFO: netdata: PROC Plugin thread created with task id -1
16-04-17 11:40:57: INFO: netdata: Successfully became user 'netdata'.
16-04-17 11:40:57: INFO: netdata: NetData started on pid 6351
16-04-17 11:40:57: INFO: netdata: Starting thread tc.
16-04-17 11:40:57: INFO: netdata: TC thread created with task id -1
16-04-17 11:40:57: INFO: netdata: Starting thread idlejitter.
16-04-17 11:40:57: INFO: netdata: CPU Idle Jitter thread created with task id -1
16-04-17 11:40:57: INFO: netdata: Starting thread proc.
16-04-17 11:40:57: INFO: netdata: PROC Plugin thread created with task id -1
16-04-17 11:41:10: INFO: netdata: Successfully became user 'BBN'.
16-04-17 11:41:10: INFO: netdata: NetData started on pid 6360
16-04-17 11:41:10: INFO: netdata: Starting thread tc.
16-04-17 11:41:10: INFO: netdata: TC thread created with task id -1
16-04-17 11:41:10: INFO: netdata: Starting thread idlejitter.
16-04-17 11:41:10: 16-04-17 11:41:10: INFO: netdata: CPU Idle Jitter thread created with task id -1INFO: netdata: 
Starting thread proc.
16-04-17 11:41:10: INFO: netdata: PROC Plugin thread created with task id -1
16-04-17 11:41:10: ERROR: netdata: Cannot create/open file '/var/cache/netdata/system.idlejitter/main.db'. (errno 13, (null))
16-04-17 11:41:10: INFO: netdata: Starting thread plugins.d.
16-04-17 11:41:10: ERROR: netdata: Cannot create/open file '/var/cache/netdata/system.idlejitter/jitter.db'. (errno 13, (null))
16-04-17 11:41:10: INFO: netdata: Not starting thread check.16-04-17 11:41:10: 
INFO: netdata: PLUGINS.D thread created with task id -1
16-04-17 11:41:10: INFO: netdata: Starting thread web.
16-04-17 11:41:10: INFO: netdata: WEB SERVER thread created with task id -1
16-04-17 11:41:10: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/apps.plugin' running on pid 6362
16-04-17 11:41:10: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/charts.d.plugin' running on pid 6363
16-04-17 11:41:10: INFO: netdata: executing command: 'exec /usr/libexec/netdata/plugins.d/apps.plugin 1 ' on pid 6362.
16-04-17 11:41:10: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/node.d.plugin' running on pid 6364
16-04-17 11:41:10: INFO: apps.plugin: Found NETDATA_HOST_PREFIX=''
16-04-17 11:41:10: INFO: apps.plugin: Found NETDATA_CONFIG_DIR='/etc/netdata'
16-04-17 11:41:10: INFO: apps.plugin: starting...
16-04-17 11:41:10: ERROR: apps.plugin: PROCFILE: Cannot open file '/proc/sys/kernel/pid_max' (errno 2, (null))
16-04-17 11:41:10: ERROR: apps.plugin: PROCFILE: Cannot open file '/proc/stat' (errno 2, (null))
16-04-17 11:41:11: ERROR: apps.plugin: Cannot collect /proc data for running processes. Disabling apps.plugin... (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/sys/kernel/mm/ksm/pages_shared' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/sys/kernel/mm/ksm/pages_sharing' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: 16-04-17 11:41:11: Cannot create/open file '/var/cache/netdata/netdata.apps_cpu/main.db'.ERROR: netdata:  (errno 13, (null))
PROCFILE: Cannot open file '/sys/kernel/mm/ksm/pages_unshared' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/sys/kernel/mm/ksm/pages_volatile' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_cpu/user.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/sys/kernel/mm/ksm/pages_to_scan' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_cpu/system.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/loadavg' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_files/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/interrupts' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_files/files.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/softirqs' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_files/pids.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/sys/kernel/random/entropy_avail' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_files/fds.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/dev' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.apps_files/targets.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/diskstats' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/apps.plugin' called DISABLE. Disabling it.
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/snmp' (errno 2, (null))
16-04-17 11:41:11: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/apps.plugin' on pid 6362 stopped.
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/snmp6' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: pid 6362 exited with code 1.
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/netstat' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/apps.plugin' (pid 6362) exited with code 1. Disabling it.
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/stat/nf_conntrack' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/ip_vs_stats' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/stat/synproxy' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/stat' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/meminfo' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/vmstat' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: PROCFILE: Cannot open file '/proc/net/rpc/nfsd' (errno 2, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.plugin_proc_cpu/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.plugin_proc_cpu/user.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.plugin_proc_cpu/system.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.server_cpu/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.server_cpu/user.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.server_cpu/system.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.clients/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.clients/clients.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.requests/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.requests/requests.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.net/main.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.net/in.db'. (errno 13, (null))
16-04-17 11:41:11: ERROR: netdata: Cannot create/open file '/var/cache/netdata/netdata.net/out.db'. (errno 13, (null))
16-04-17 11:41:12: ERROR: netdata: 1: File '/usr/share/netdata/web/' is owned by user 503 (expected user 501). Access Denied. (errno 3, (null))
16-04-17 11:41:13: ERROR: netdata: 1: File '/usr/share/netdata/web/favicon.ico' is owned by user 503 (expected user 501). Access Denied.
16-04-17 11:41:20: ERROR: netdata: pid 6361 killed by signal 0. (errno 3, (null))
16-04-17 11:41:20: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/node.d.plugin' on pid 6364 stopped.
16-04-17 11:41:20: INFO: netdata: PLUGINSD: '/usr/libexec/netdata/plugins.d/charts.d.plugin' on pid 6363 stopped.
16-04-17 11:41:20: INFO: netdata: Signaled exit (signal 15). (errno 60, (null))
16-04-17 11:41:20: ERROR: netdata: pid 6363 killed by signal 0.
16-04-17 11:41:20: ERROR: netdata: pid 6364 killed by signal 0.
16-04-17 11:41:20: INFO: netdata: NetData exiting. Bye bye...
16-04-17 11:41:26: INFO: netdata: Successfully became user 'netdata'.
16-04-17 11:41:26: INFO: netdata: NetData started on pid 6370
16-04-17 11:41:26: INFO: netdata: Starting thread tc.
16-04-17 11:41:26: INFO: netdata: Starting thread idlejitter.
16-04-17 11:41:26: INFO: netdata: TC thread created with task id -1
16-04-17 11:41:26: INFO: netdata: CPU Idle Jitter thread created with task id -1
16-04-17 11:41:26: INFO: netdata: Starting thread proc.
 ```

**Comment by ktsaou** [comment]
> 16-04-17 11:41:12: ERROR: netdata: 1: File '/usr/share/netdata/web/' is owned by user 503 (expected user 501). Access Denied. (errno 3, (null))
> 16-04-17 11:41:13: ERROR: netdata: 1: File '/usr/share/netdata/web/favicon.ico' is owned by user 503 (expected user 501). Access Denied.

These are the correct chowns.

All the errors about files in `/proc/` and `/sys/` are about linux specific data collection. They will not work on OS X. So, even if you manage to start it, I am not sure it will show any charts.

**Comment by Yanbo** [comment]
Please make a OSX version~~~This looks really awesome.

**Comment by hellobbn** [comment]
Is the function of netdata the same as "glances" ? 
Glances can work on mac quit well.

**Comment by deltadelphini** [comment]
+1 requesting an osx version, please

**Comment by leewi9** [comment]
+1 please make our dream come true asap


2016年5月26日 +0800 15:27 deltadelphini<notifications@github.com>，写道：

**Comment by ktsaou** [comment]
You are so nice...

**Comment by sammcj** [comment]
+1 This would be _amazing_, I'm no developer but happy to test and provide feedback if you need people to do so.

**Comment by Eggwise** [comment]
Hi.
Just got a macbook pro (newest version) and **i would love** to use netdata.
Any experiences running netdata on mac? is it safe to install it?
cheers

**Comment by hellobbn** [comment]
Noticed a guide on how to install GitLab on OS X [here](https://github.com/WebEntity/Installation-guide-for-GitLab-on-OS-X). I think this may be put into use.

**Comment by kiela** [comment]
@hellobbn And how GitLab is related to this project because I don't get it?

**Comment by hellobbn** [comment]
For I am not actually aware of how it works, maybe I reached the wrong conclusion? Sorry...... @kiela

**Comment by MiracleBlue** [comment]
Oh gracious and wise @ktsaou, we humbly request the presence of your genius to grace our Mac hardware, so that we simple folk may also bask forever in your glorious charts and graphs, although we are not worthy!  
In your honour, I have bought forth this offering, may it please you:  🐐 (that's supposed to be a goat, and its supposed to be bigger... but this is the best I could do)

Hyperbole aside, I seriously would love to see a tool as gorgeous and detailed as this available for Mac, just for my own use, cause there's really nothing anywhere near as detailed as what I saw in the Netcat demos (I am definitely open to suggestions if anyone knows of any!)

tldr: Plz firehol ༼ つ ◕_◕ ༽つ give NETCAT

**Comment by ktsaou** [comment]
:)

Thanks to the wonderful work of @vlvkobal netdata (not netcat), works on FreeBSD. Since MacOS X has its roots on BSDs, have you tried to install it on Mac OS X? Check https://github.com/firehol/netdata/pull/1321 and https://github.com/firehol/netdata/wiki/Installation#freebsd

For sure, you need to convince someone with the proper skills to port and maintain netdata on Mac OS X. I will help him/her, but I cannot do all work myself. Given that netdata is still in its early development phases, I prefer to focus on how to make netdata, than port the existing features to every system on planet Earth.

So, I just need help... (can you take care of something for netdata? check this: https://github.com/firehol/netdata/wiki#can-i-help)

**Comment by MiracleBlue** [comment]
I'd love to help, if I ever find the time outside of all of my own projects also :P I thought the above commenters with the error logs were attempting to install it on OS X (and getting all the above errors) so I assumed it wouldnt not work.  I'm on a brand new MBP w/ Touch Bar and Sierra that I can test installation on, so perhaps I'll give that a try and maybe it will work just like that!  I will report back if I get a chance.

Regardless, I wish you continued success, you're clearly onto a good thing here :)

**Comment by MiracleBlue** [comment]
Also how did I end up calling it netcat in my original comment I don't know... I must've been tired!

**Comment by ktsaou** [comment]
> I thought the above commenters with the error logs were attempting to install it on OS X (and getting all the above errors) so I assumed it wouldnt not work.

It is a timing issue. @vlvkobal did his work in last 2 weeks...

**Comment by MiracleBlue** [comment]
Ah-ha!  Well that explains everything.  I'm one of those awful speed readers that sometimes rushes through pages without fully making sure I have all the information... 

Appreciate it, I will give this a test on Sierra sometime soon and report back!

**Comment by ktsaou** [comment]
As of today netdata runs on Max OS X, thanks to @vlvkobal for PR #1419.
It still has limited charts, but external plugins do work.

**Comment by vlvkobal** [comment]
Information about `netdata` installation on macOS is available at [https://github.com/firehol/netdata/wiki/Installation#macos](https://github.com/firehol/netdata/wiki/Installation#macos)

**Comment by craigmcdonnell** [comment]
FYI, I tried installing netdata on Mac OS X (El Capitan v.10.11.6) today and it failed when attempting to chown on ~/opt/netdata/usr/libexec/netdata/plugins.d/apps.plugin. Creating an empty file at the specified path and then re-running the installer allowed installation to complete.

**Comment by simonnagl** [comment]
@craigmcdonnell I tried to reproduce this on macOS Sierra (Version 10.12.2). There the installer just complains. This is OK because we currently do not compile apps.plugins. It will not work either because it needs to be ported to macOS. This is not done yet. So if it just warns you might just ignore this.

**Comment by craigmcdonnell** [comment]
It failed to install until I created ~/opt/netdata/usr/libexec/netdata/plugins.d/apps.plugin. Also, worth noting, I couldn't install netdata as normal user so had to use sudo. Having installed in this manner, I then needed to chown of all the installed netdata files to current user in order to access the netdata site.

**Comment by sebkim** [comment]
@craigmcdonnell I have also exactly same issue.

**Comment by ktsaou** [comment]
If anyone is willing to make a PR, please do so... I don't have access to these versions of MacOS.

**Comment by simonnagl** [comment]
Open #1789 to track the bug there. There it is more likely found.

**Comment by ktsaou** [comment]
Cleaning up. MacOS support is here, so if you have any issues with it, please open a new issue.
